eng/code-quality-reports/src/main/java/com/azure/tools/checkstyles/checks/ServiceClientChecks.java

```diff
@@ -0,0 +1,152 @@
+    }
+
+    @Override
+    public int[] getAcceptableTokens() {
+        return getRequiredTokens();
+    }
+
+    /**
+     * Denotes that the PACKAGE_DEF token type is the only one that should trigger visitToken to be called.
+     * This is done to prevent inner classes from accidentally firing checks.
+     * @return AST token types that will trigger visitToken to be called.
+     */
+    @Override
+    public int[] getRequiredTokens() {
+        return new int[] { TokenTypes.PACKAGE_DEF };
+    }
+
+    /**
+     * Runs the check logic on the tokens we marked as needing to be checked in getRequiredTokens.
+     * @param ast Abstract syntax tree node that will be checked.
+     */
+    @Override
+    public void visitToken(DetailAST ast) {
+        DetailAST classDefNode = CustomCheckUtils.findNextSiblingOfType(ast, (node) -> node.getType() == TokenTypes.CLASS_DEF);
+        if (classDefNode == null) {
+            return;
+        }
+
+        DetailAST objBlockNode = classDefNode.findFirstToken(TokenTypes.OBJBLOCK);
+        Optional<String> fullClassName = getFullClassName(ast, classDefNode);
+        if (!fullClassName.isPresent()) {
+            return;
+        }
+
+        // Attempt to load the class and run the checks if it is a descendant of ServiceClient.
+        try {
+            Class<?> classToCheck = this.getClassLoader().loadClass(fullClassName.get());
+            Class<?> serviceClientClass = this.getClassLoader().loadClass(ServiceClient.class.getName());
+
+            if (!serviceClientClass.isAssignableFrom(classToCheck)) {
+                return;
+            }
+        } catch (ClassNotFoundException ex) {
+            return;
+        }
+
+        TokenUtil.forEachChild(objBlockNode, TokenTypes.CTOR_DEF, this::checkConstructorIsHidden);
+        if (!TokenUtil.findFirstTokenByPredicate(objBlockNode, this::isMethodStaticBuilder).isPresent()) {
+            log(classDefNode, BUILDER_ERROR_MESSAGE);
+        }
+    }
+
+    /**
+     * Given the package node and the class definition node construct the full class name.
+     * @param packageDefNode Node containing the package name
+     * @param classDefNode Node containing the class definition
+     * @return Full class name, if the file isn't a class an empty optional
+     */
+    private Optional<String> getFullClassName(DetailAST packageDefNode, DetailAST classDefNode) {
+        String packageName = FullIdent.createFullIdent(packageDefNode.findFirstToken(TokenTypes.DOT)).getText();
+
+        // If the package begins with com.microsoft.* it is still in track one and should skip this check.
+        if (packageName.startsWith("com.microsoft")) {
+            return Optional.empty();
+        }
+
+        DetailAST classIdentifierNode = classDefNode.findFirstToken(TokenTypes.LITERAL_CLASS);
+        DetailAST classNameNode = CustomCheckUtils.findNextSiblingOfType(classIdentifierNode, (node) -> node.getType() == TokenTypes.IDENT);
```

Comment:
nit: `(node)` could be `node` since there is only one arg into this Consumer. there are a few other instances.